src/lib/rtl/rtl_time_estimator.cpp

```diff
@@ -0,0 +1,267 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file rtl_time_estimator.cpp
+ *
+ * Helper class to calculate the remaining time estimate to go to RTL landing point.
+ *
+ */
+
+#include "rtl_time_estimator.h"
+
+#include <float.h>
+#include <math.h>
+
+#include <px4_platform_common/defines.h>
+#include <drivers/drv_hrt.h>
+#include <uORB/topics/vehicle_status.h>
+
+RtlTimeEstimator::RtlTimeEstimator() : ModuleParams(nullptr)
+{
+	_param_mpc_z_v_auto_up = param_find("MPC_Z_V_AUTO_UP");
+	_param_mpc_z_v_auto_dn = param_find("MPC_Z_V_AUTO_DN");
+	_param_mpc_land_speed = param_find("MPC_LAND_SPEED");
+	_param_fw_climb_rate = param_find("FW_T_CLMB_R_SP");
+	_param_fw_sink_rate = param_find("FW_T_SINK_R_SP");
+	_param_fw_airspeed_trim = param_find("FW_AIRSPD_TRIM");
+	_param_mpc_xy_cruise = param_find("MPC_XY_CRUISE");
+	_param_rover_cruise_speed = param_find("GND_SPEED_THR_SC");
+};
+
+rtl_time_estimate_s RtlTimeEstimator::getEstimate() const
+{
+	rtl_time_estimate_s time_estimate{};
+
+	if (_is_valid && PX4_ISFINITE(_time_estimate)) {
+		time_estimate.valid = true;
+		time_estimate.time_estimate = _time_estimate;
+		// Use actual time estimate to compute the safer time estimate with additional scale factor and a margin
+		time_estimate.safe_time_estimate = _param_rtl_time_factor.get() * _time_estimate + _param_rtl_time_margin.get();
+
+	} else {
+		time_estimate.valid = false;
+	}
+
+	time_estimate.timestamp = hrt_absolute_time();
+	return time_estimate;
+}
+
+void RtlTimeEstimator::update()
+{
+	_vehicle_status_sub.update();
+	_wind_sub.update();
+
+	if (_parameter_update_sub.updated()) {
+		parameter_update_s param_update;
+		_parameter_update_sub.copy(&param_update);
+
+		// If any parameter updated, call updateParams() to check if
+		// this class attributes need updating (and do so).
+		updateParams();
+	}
+}
+
+void RtlTimeEstimator::addVertDistance(float alt)
+{
+	if (PX4_ISFINITE(alt)) {
+		_is_valid = true;
+
+		_time_estimate += calcVertTimeEstimate(alt);
+	}
+}
+
+void RtlTimeEstimator::addDistance(float hor_dist, const matrix::Vector2f &direction, float vert_dist)
+{
+	if (PX4_ISFINITE(hor_dist) && PX4_ISFINITE(vert_dist)) {
+		_is_valid = true;
+
+		float hor_time_estimate{0.f};
+
+		if (hor_dist > FLT_EPSILON && PX4_ISFINITE(hor_dist)) {
+			hor_time_estimate = hor_dist / getCruiseGroundSpeed(direction.normalized());
+		}
+
+		float ver_time_estimate{calcVertTimeEstimate(vert_dist)};
+
+		_time_estimate += math::max(hor_time_estimate, ver_time_estimate);
```

Comment:
This could be improved - if you have similar altitude and horizontal differences then both matter, not the max().

Reply:
How do you mean that?